keps/sig-storage/177-volume-snapshot/tighten-validation-webhook-crd.md

```diff
@@ -0,0 +1,815 @@
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingWebhookConfiguration
+metadata:
+  name: "webhook-validation.storage.sigs.k8s.io"
+webhooks:
+- name: "snapshot.webhook-validation.storage.sigs.k8s.io"
+  rules:
+  - apiGroups:   ["snapshot.storage.k8s.io"]
+    apiVersions: ["v1beta1"]
+    operations:  ["CREATE", "UPDATE"]
+    resources:   ["volumesnapshots", "volumesnapshotcontents"]
+    scope:       "*"
+  clientConfig:
+    service:
+      namespace: "default"
+      name: "snapshot-validation-service"
+      path: "/path/to/webhook"
+    caBundle: "LS0tLS...base64 encoded of public key...LS0K"
+  admissionReviewVersions: ["v1", "v1beta1"]
+  sideEffects: None
+  failurePolicy: Ignore # We recommend switching to Fail only after successful installation of the server and webhook.
+  timeoutSeconds: 2 # This will affect the latency and performance. Finetune this value based on your application's tolerance.
+```
+
+Webhook Server Deployment
+The recommended deployment mode for the webhook server is within the cluster to minimize network latency. For high-availability we recommend using a Deployment and Service to deploy the validation server. Some example yaml files are provided, and should be changed to suit the Cluster Admin’s needs.
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: snapshot-validation-deployment
+  labels:
+    app: snapshot-validation
+spec:
+  replicas: 3
+  selector:
+    matchLabels:
+      app: snapshot-validation
+  template:
+    metadata:
+      labels:
+        app: snapshot-validation
+    spec:
+      containers:
+      - name: snapshot-validation
+        image: image:xxx # change the image to released image or if you wish to use your own custom validation server image
+        args: ['webhook', '--tls-cert-file=/etc/webhook/certs/cert.pem', '--tls-private-key-file=/etc/webhook/certs/key.pem'] # Change args as needed
+        ports:
+        - containerPort: 443 # change the port as needed
+        volumeMounts:
+          - name: webhook-certs
+            mountPath: /etc/webhook/certs
+            readOnly: true
+      volumes:
+        - name: webhook-certs
+          secret:
+            secretName: snapshot-validation-secret
+```
+
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: snapshot-validation-service
+  namespace: default # Don't use the default namespace. Choose an appropriate one.
+spec:
+  selector:
+    app: snapshot-validation
+  ports:
+    - protocol: TCP
+      port: 443 # Change if needed
+      targetPort: 443 # Change if the webserver image expects a different port
+```
+
+### Test Plan
+There will be unit testing on the webserver to ensure that the correct policy gets enforced.
+
+There is a cluster available for testing in the external snapshotter repository. We will deploy the webhook validation server and test the e2e functionality.
```

Comment:
"There is a cluster available for testing in the external snapshotter repository."
Do you mean we can run CI from the external snapshotter repository which sets up a k8s cluster?

Reply:
It may be sufficient to say that the kubernetes-csi and k/k e2e tests will be enhanced.

Reply:
We discussed offline. Since the webhook is developed in external-snapshotter and does not test any csi driver, it may be better to add a new test job in external-snapshotter that brings up a kind cluster, installs the crds and then run specific validation tests.

Reply:
Added a note on testing.